### PR TITLE
Implemented option to Hide the "spill" bucket from categorical histograms behind an advanced setting

### PR DESCRIPTION
**Pull request by ab0092**

### CODE REVIEW

File: new-pipeline/dist.html [hunk 1]
```diff
@@ -188,6 +188,10 @@ <h3><select id="selected-key4" class="multiselect" title="Selected key"></select
                                     <label><input name="trim-toggle" type="checkbox"> <b>Trim small buckets</b> </label>
                                     Hide buckets on both sides that don't have enough samples.
                                 </div>
+                                <div>
+                                    <label><input name="include-spill-toggle" type="checkbox"> <b>Include Spill</b> </label>
+                                    Include Spill Data in Results
```
Comment (chutten) [inline new-pipeline/dist.html]: Let's go with "Include 'spill' bucket (for categorical histograms)"

File: new-pipeline/evo.html [hunk 2]
```diff
@@ -97,6 +97,10 @@ <h2 id="sample-counts-title" style="font-size: 18px; margin-bottom: 0"></h2>
                                     <label><input name="sanitize-toggle" type="checkbox"> <b>Sanitize data</b> </label>
                                     Hide data points that don't have enough submissions.
                                 </div>
+                                <div>
+                                    <label><input name="include-spill-toggle" type="checkbox"> <b>Include Spill</b> </label>
+                                    Include Spill Data in Results
```
Comment (chutten) [inline new-pipeline/evo.html]: Same here: "Include 'spill' bucket (for categorical histograms)"

File: new-pipeline/src/dist.js [hunk 3]
```diff
@@ -250,7 +253,6 @@ $(function () {
                 if (getAggregate === undefined) {
                   throw "Could not obtain aggregate function"
                 };
-
```
Comment (chutten) [inline new-pipeline/src/dist.js:253]: Please put this line back in.

File: new-pipeline/src/dist.js [hunk 4]
```diff
@@ -788,10 +791,19 @@ function updateDateRange(callback, dates, updatedByUser, shouldUpdateRangebar) {
   }
 }
 
-function displayHistograms(histogramsList, dates, useTable, cumulative, trim) {
+function displayHistograms(histogramsList, dates, useTable, cumulative, trim, includeSpill) {
   cumulative = cumulative === undefined ? false : cumulative;
   trim = trim === undefined ? true : trim;
 
+  if(!includeSpill)// Hide the 'spill' related data
```
Comment (chutten) [inline new-pipeline/src/dist.js]: Some style comments: 

The comment should be on its own line, above the code it is commenting, and be a complete sentence (starting with a capital letter, ending with a period)
The opening brace should be on the same line as the if
The keyword if should have a space after it.

```patch
-if(!includeSpill)// Hide the 'spill' related data
-{
+// Hide the 'spill' bucket if present.
+if (!includeSpill) {
```

File: new-pipeline/src/dist.js [hunk 5]
```diff
@@ -1870,3 +1884,24 @@ function saveStateToUrlAndCookie() {
 function isHistogramsListKeyed(histogramsList) {
   return histogramsList.some(entry => entry.title !== "");
 }
+
+function makenewHistogramListCopy(histogramlist) {
+ return histogramlist.map((item) => {
+    return {
+      title: item.title,
+      histograms: item.histograms.map((histogram) => {
```
Comment (chutten) [inline new-pipeline/src/dist.js]: I think we can use some ECMA6 notation here to save us some trouble:

```js
return histogramList.map(item => {
  return {
    ...item,
    histograms: item.histograms.map(histogram => {
      return {
        ...histogram,
        buckets: histogram.buckets.slice(0),
        values: histogram.values.slice(0),
      };
    }),
  };
});
```

This is javascript's [spread syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax), and it comes in handy all over the place.

File: new-pipeline/src/dist.js [hunk 6]
```diff
@@ -788,10 +791,19 @@ function updateDateRange(callback, dates, updatedByUser, shouldUpdateRangebar) {
...
 }
 
-function displayHistograms(histogramsList, dates, useTable, cumulative, trim) {
+function displayHistograms(histogramsList, dates, useTable, cumulative, trim, includeSpill) {
   cumulative = cumulative === undefined ? false : cumulative;
   trim = trim === undefined ? true : trim;
 
+  if(!includeSpill)// Hide the 'spill' related data
+  {
+    const spillindex = histogramsList[0].histograms[0].buckets.indexOf("spill");
```
Comment (chutten) [inline new-pipeline/src/dist.js]: use `camelCase` for identifiers, please. (this one would be `spillIndex`)

File: new-pipeline/src/dist.js [hunk 7]
```diff
@@ -1870,3 +1884,24 @@ function saveStateToUrlAndCookie() {
 function isHistogramsListKeyed(histogramsList) {
   return histogramsList.some(entry => entry.title !== "");
 }
+
+function makenewHistogramListCopy(histogramlist) {
+ return histogramlist.map((item) => {
```
Comment (chutten) [inline new-pipeline/src/dist.js]: Please use `camelCase` for identifiers. This one would be `histogramList`

**Review by chutten — COMMENTED** [review]
It behaves very well under my testing, this is great

**Comment by ab0092** [comment]
@chutten i have done the required changes, but we have a problem if we use the spread operator while creating a new copy of the **histogramList** because each Histogram has a **map** function attached to its prototype property that is why i have used **Object.Create** using the Histogram's prototype or else the code will not work as expected. Please let me know if we can come up with a workaround or we are going to keep that code section unchanged. Thanks..

**Comment by chutten** [comment]
Ah yes, and that would be used during the conversion to data arrays for metrics graphics to display. That makes sense, and makes a shallow copy not work. We're fine without changing that section.

**Review by chutten — COMMENTED** [review]
It seems as though you changed the style of the whole of dist.js and that makes it difficult to review. Can you limit the changes to the meaningful code changes for now and we can consider formatting the code in a different PR?

**Comment by ab0092** [comment]
I made a mistake with my editor and ended up getting the whole dist.js formatted. Any idea how i can roll it back. I have already commited that.

**Comment by chutten** [comment]
You can try `git revert HEAD` which will make a commit that undoes everything in the `HEAD` commit. That'll get your branch back to the state where it was before your most recent commit, but it'll leave your commits on the branch so you can look at them (with `git show HEAD^` for instance).

If you don't want to leave the commits on the branch you can `git reset --hard HEAD^` which will reset your branch back to the state it was before the most recent commit by deleting your most recent commit from history.

**Comment by ab0092** [comment]
Thanks for the help. I rolled it back and committed the required changes.

File: new-pipeline/src/dist.js [hunk 8]
```diff
@@ -1640,6 +1651,8 @@ function saveStateToUrlAndCookie() {
       .format("YYYY-MM-DD"),
     end_date: moment(picker.endDate)
       .format("YYYY-MM-DD"),
+    include_spill: $("input[name=include-spill-toggle]")
+    .is(":checked") ? 1 : 0
```
Comment (chutten) [inline new-pipeline/src/dist.js]: Need a level of indentation here (two spaces)

File: new-pipeline/src/evo.js [hunk 9]
```diff
@@ -394,8 +397,12 @@ var gLoadedAggregatesFromState = false;
 function updateAggregates(kind, buckets) {
   gCurrentKind = kind;
   let newAggregates;
+  const IncludeSpill = $("input[name=include-spill-toggle]").is(":checked");
+
   if (kind === "enumerated" || kind === "categorical") {
     newAggregates = getHumanReadableBucketOptions(kind, buckets);
+    // Hide spill Option from Aggregates based on advanced settings
+    if(!IncludeSpill) newAggregates = newAggregates.filter(([key,value]) => value !== "spill");
```
Comment (chutten) [inline new-pipeline/src/evo.js]: Conditionals should be a minimum of three lines
```patch
-if (cond) statement();
+if (cond) {
+  statement();
+}
```

(also, space required between `if` and `(`)

File: new-pipeline/src/evo.js [hunk 10]
```diff
@@ -944,6 +950,8 @@ function saveStateToUrlAndCookie() {
       .is(":checked") ? 1 : 0,
     sanitize: $("input[name=sanitize-toggle]")
       .is(":checked") ? 1 : 0,
+      include_spill: $("input[name=include-spill-toggle]")
```
Comment (chutten) [inline new-pipeline/src/evo.js]: Please back this line out a level of indent (2 spaces) to line up with `sanitize`

File: new-pipeline/src/dist.js [hunk 11]
```diff
@@ -788,9 +791,17 @@ function updateDateRange(callback, dates, updatedByUser, shouldUpdateRangebar) {
   }
 }
 
-function displayHistograms(histogramsList, dates, useTable, cumulative, trim) {
+function displayHistograms(histogramsList, dates, useTable, cumulative, trim, includeSpill) {
   cumulative = cumulative === undefined ? false : cumulative;
   trim = trim === undefined ? true : trim;
+  // Hide the 'spill' bucket if present
+  if(!includeSpill) {
```
Comment (chutten) [inline new-pipeline/src/dist.js]: need a space between `if` and `(`

File: new-pipeline/src/dist.js (hunk 3 above)
Comment (chutten) [reply new-pipeline/src/dist.js:253]: Still looking for this line : )

**Review by chutten — COMMENTED** [review]
Nearly there!

**Comment by ab0092** [comment]
I think these changes this should do it. :-)

File: new-pipeline/src/dist.js [hunk 12]
```diff
@@ -251,7 +251,7 @@ $(function () {
                 }[$("#sort-keys")
                   .val()];
                 if (getAggregate === undefined) {
-                  throw "Could not obtain aggregate function"
+                    throw "Could not obtain aggregate function"
```
Comment (chutten) [inline new-pipeline/src/dist.js]: This seems unintentional

**Review by chutten — COMMENTED** [review]
Ack, it seems as though an extra change snuck in

**Comment by ab0092** [comment]
Ya i some how i added the extra indent.

**Comment by ab0092** [comment]
As for the code on line no **253 - 255** in **dist.js** the code already follows an indentation with the rest of the code. Do you want any changes there??

**Comment by chutten** [comment]
It isn't the indentation in this case. In this case you deleted a line that I think might reduce the readability of the code. (whitespace is helpful)

**Comment by ab0092** [comment]
Is there anything else that i am missing here or are we good to merge this ???

**Comment by chutten** [comment]
That's it, that's all, your code is now merged!

**Comment by ab0092** [comment]
Thanks.. And sorry for the slight delay in completing the formatting changes.. :-)

**Pull request merged**